feat: implement vim.secure.read() and use it for 'exrc'

gpanders opened the pull request:
Add the `vim.secure.read()` function in Lua which prompts the user whether the given file is trusted or not. This function is used to implement the `'exrc'` feature. By explicitly asking the user whether or not a `.nvimrc` or `.exrc` file is trusted (and only sourcing the file if so), the most trivial security issues with the `'exrc'` options are resolved and the function can be un-deprecated.

Closes #20911.

gpanders committed:
feat: add vim.secure.read()

This function accepts a path to a file and prompts the user if the file
is trusted. If the user confirms that the file is trusted, the contents
of the file are returned. The user's decision is stored in a trust
database at $XDG_STATE_HOME/nvim/trust. When this function is invoked
with a path that is already marked as trusted in the trust database, the
user is not prompted for a response.

gpanders committed:
feat(exrc): use vim.secure.read() for 'exrc' option

ii14 commented on runtime/lua/vim/_editor.lua:
```diff
@@ -892,6 +892,108 @@ function vim._init_default_mappings()
+
+    vim.api.nvim_buf_set_lines(buf, 0, -1, false, lines)
+    vim.cmd('new')
+    local win = vim.api.nvim_get_current_win()
+    vim.api.nvim_win_set_buf(win, buf)
```
```suggestion
    vim.api.nvim_set_current_buf(buf)
```

gpanders replied:
Fixed in a1dc757291aed157943cc0ac94a634acc2f0d40c

ii14 commented on src/nvim/lua/executor.c:
```diff
@@ -2177,3 +2177,14 @@ char *nlua_funcref_str(LuaRef ref)
+  lua_pushstring(lstate, path);
+  lua_newtable(lstate);
+  lua_call(lstate, 2, 1);
+  return lua_tostring(lstate, -1);
```
`vim` and the returned string is not popped from the stack. Popping the string from the stack could free the string, so I think it just has to be copied. `lua_tolstring` will also return the length, that can be used for malloc and memcpy

gpanders replied:
Fixed in 4a2674d4effcdb7a742926000cc9de6d3522c3f0

gpanders commented on runtime/lua/vim/_editor.lua:
```diff
@@ -892,6 +892,108 @@ function vim._init_default_mappings()
+--- $XDG_STATE_HOME/nvim/trust.
+---
+---@param path (string) Path to a file to read.
+---@param opts (table|nil) Options (currently unused).
```
Fixed in 864ad2711dd8a1955b6839d3d17108504a129488

gpanders commented on runtime/lua/vim/_editor.lua:
```diff
@@ -892,6 +892,108 @@ function vim._init_default_mappings()
+---
+---@return (string|nil) The contents of the given file if it is trusted, or nil
+---                     otherwise.
+function vim.readsecure(path, opts)
```
41d591ab56946f8cb5a3b9db60013a74ae1f1da2

ii14 commented on runtime/lua/vim/secure.lua:
```diff
@@ -0,0 +1,104 @@
+    end
+
+    vim.api.nvim_buf_set_lines(buf, 0, -1, false, lines)
+    vim.cmd('new')
```
`:new` creates a new buffer, it can replace `nvim_create_buf` entirely.

gpanders replied:
Fixed in 61cea11c82492c90c6b465deceaf9a53b3851c04

ii14 commented on src/nvim/lua/executor.c:
```diff
@@ -2177,3 +2177,26 @@ char *nlua_funcref_str(LuaRef ref)
+  const char *contents = lua_tolstring(lstate, -1, &len);
+  char *buf = NULL;
+  if (contents != NULL) {
+    buf = xcalloc(len, sizeof(char));
```
I'm not sure about this, but I think `len` might not include the null byte.

gpanders replied:
It does not. Fixed in d1a847ca9a7323db1b74a2fd77ee8b351be1696d

gpanders commented on runtime/lua/vim/secure.lua:
```diff
@@ -0,0 +1,102 @@
+      local contents = f:read('*a')
+      if contents then
+        for line in vim.gsplit(contents, '\n') do
+          local hash, file = string.match(line, '^(%S+) (.+)$')
```
Is this actually necessary? The input path argument is already expanded to a full (absolute) path, so if a relative path exists here it would never compare equal against the absolute input path.

ii14 replied:
Yes, you're right, it's not necessary. What worries me however is that `:p`'s behavior in `fnamemodify` is "unpredictable" if file doesn't exist, and that could potentially lead to something unexpected. Can you replace that with `fs_realpath`? `fs_realpath` fails when file doesn't exist

gpanders replied:
Fixed in 7578fc69f2f9940b36b019a7d3fe0729302890e6. I think it might make sense to also use `fs_realpath` in `vim.fs.normalize()`. If we make that change then we can remove it here.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2353,4 +2353,20 @@ parents({start})                                            *vim.fs.parents()*
 
+
+==============================================================================
+Lua module: secure                                                *lua-secure*
```
I see that `vim.readsecure()` https://github.com/neovim/neovim/pull/20925 was closed but don't see a rationale. Why introduce a new module for this? Do we plan to have many more functions in `vim.secure`? 

The current functionality at least seems to be covered by a parameter on `vim.read({secure=true, path=...})`. Otherwise `vim.readsecure()` works, if the interface is expected to be different than `vim.read()`.

gpanders replied:
See @ii14's [comment](https://github.com/neovim/neovim/pull/20956#discussion_r1014756332). The thinking was there may be other useful APIs that could fall under the "secure" umbrella (e.g. an interface to add or remove items from the trust database).

justinmk replied:
"vim.secure" is quite a lofty name for this :) Maybe `vim.localrc` or `vim.exrc`. 

Not trying to bikeshed, but thinking of `trust()` and looking at https://github.com/ii14/exrc.vim/blob/master/doc/exrc.txt and related plugins, clearly the direction is to support the features of these "local vimrc". `vim.secure` isn't the right name for this concept.

ii14 replied:
vimrcs are the most obvious use case for this, but it could be used for anything really, not necessarily even vim script or lua.

gpanders replied:
>The current functionality at least seems to be covered by a parameter on vim.read({secure=true, path=...}). Otherwise vim.readsecure() works, if the interface is expected to be different than vim.read().

I do think that the "readsecure" interface might be different enough to warrant being separate from a more generic `read` function. For example, we might want to expose the ability to specify the prompt text, or modify some parameters of how the file is displayed when the user wants to view it (e.g. setting the buffer filetype). None of these would make sense for a generic `read` function.

Re: `vim.secure.read()` vs `vim.readsecure()`: I don't have a dog in that fight, but I'd prefer we just pick something and not bikeshed too long. It sounds like @ii14 already has some ideas on other functions that would fit in a "secure" module. We also have plenty of time until the next release cycle if we want to just throw it under `vim.readsecure()` for now and refactor into a separate module later on when we have more use cases.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2253,6 +2253,18 @@ A jump table for the options with a short description can be found at |Q_op|.
+	directory.
+
+	The file is only sourced if the user indicates the file is trusted.
+	See |vim.secure.read()|.
```
The hash check is worth mentioning here too.

gpanders replied:
33c1e7ec33d3502bd779849a89a8411275438e1f

justinmk commented:
Need to update the E12 this message, it's [confusing](https://github.com/neovim/neovim/issues/9881) and after this PR will be wrong(?): https://github.com/neovim/neovim/blob/16eed6e7663afccfe36fd8c0301927fd1858c5c8/src/nvim/globals.h#L866

gpanders committed:
refactor: deprecate 'secure' option

Now that 'exrc' files must be explicitly marked trusted there is no need
to constrain what can be done in them.

gpanders commented on src/nvim/main.c:
```diff
@@ -1987,7 +1987,11 @@ static void source_startup_scripts(const mparm_T *const parmp)
+        xfree(str);
+      } else {
 #if defined(UNIX)
         // if ".exrc" is not owned by user set 'secure' mode
```
Yes, and there is a lot of other code we can remove as well because of this. eabc7a5dadcd25821a1b0645681e87e118d075a2

gpanders commented:
>E12 won't be used anymore, since we don't need to set secure when sourcing a .nvimrc or .exrc file.

Actually I'm going to walk this back a bit. There are still cases where we can be in a "secure" mode, but sourcing a .nvimrc or .exrc file will not enable it. So I've updated the message to `E12: Command not allowed in secure mode in current dir or tag search`.

ii14 commented:
Looks good to me. We're going to need some way of having lua files, but even before we add that it's going to be few lines of code to implement it on your own after this PR.

gpanders commented:
Rebased on master. I am going to go ahead and merge when CI passes just so we can move forward. We can eliminate the `vim.secure` module and rename it back to `vim.readsecure()` if it turns out we don't need a full module, but no reason we can't do that in a follow up.

justinmk commented:
Sorry for the delay! Something else I forgot to mention: this is a step towards the _project_ concept.

- https://github.com/neovim/neovim/issues/8610

That further suggests that `vim.secure` is not the central topic for this idea. Perhaps `vim.config`? `vim.workspace`? Or do these features actually belong in `vim.fs` ?

ii14 commented:
@justinmk I'd frame it as maybe a building block for projects, not necessarily a part of it. The way I personally imagine projects is similar to lspconfig - you set up some pattern for the project and match buffers against it. At least in my idea of how projects would work, there is nothing that enforces where these patterns are supposed to be defined. So you could use `vim.secure` to scan for project files, or you could just as well define them in init.vim/lua.

 It would be nice to have some other functions here, namely for whitelisting and blacklisting files (`vim.secure.{allow,deny}`), that's why I thought it would be nice to have it be its own module. But I suppose it could be a submodule of `vim.fs` like `vim.fs.secure.read()` or something similar instead.

gpanders commented:
>Or do these features actually belong in vim.fs ?

I think a `vim.fs.read()` function with a `secure` option could make sense, but I think we also want an API to manage the trust file (something like `vim.secure.allow()/deny()` like @ii14 mentioned), and that does not seem like a good fit for `vim.fs`.

kristijanhusak commented:
Is loading `.exrc/.nvimrc` file intended to work only with `init.vim`, or it should also support `init.lua` ?
From what I see in https://github.com/neovim/neovim/blob/master/src/nvim/main.c#L1913-L1934, this currently works only if user is using `init.vim` as a configuration file.

gpanders commented:
It should also work with init.lua. I am not sure why that function is written that way. Please open an issue (or better yet, a PR).

kristijanhusak commented:
>  I am not sure why that function is written that way

I opened a similar issue some time ago https://github.com/neovim/neovim/issues/13501 and this comment adds some reasoning around it https://github.com/neovim/neovim/issues/13501#issuecomment-755326024

justinmk commented:
> But I suppose it could be a submodule of `vim.fs` like `vim.fs.secure.read()` or something similar instead.

Not in favor of a sub-submodule. Before adding a module ask if it is really a major topic or if it can be functions on an existing module. Over-granularity defeats the purpose of having modules because having 100 modules is no better than having a big flat list of functions: user has no idea where anything lives. Sub-sub-modules are even worse because now you have to drill into a tree to find things.
